### PR TITLE
Implement measurement pattern for Toffoli gate

### LINKED ISSUE #49: Measurement pattern for Toffoli gate
Implement measurement pattern for Toffoli gate in `graphix.Circuit` class for transpilation. Please consider:
- minimize the number of commands by choosing an appropriate decomposition before translating into a measurement pattern.
- Make sure the `Circuit.simulate_statevector()` runs with Toffoli gate.
- Please also implement 'opt' version (see, e.g. `Circuit._rz_command_opt` and `_rzz_command_opt):
    -  By performing Pauli measurement preprocessing for gate-level measurement patterns (eg `Circuit._rz_command`), it is possible to implement a significantly shorter measurement pattern for a gate, with the cost of added local Clifford commands (see `Circuit._rz_command_opt`).
    - Make sure you only perform Pauli measurements on the intermediate qubits (not input or output qubits of the gate-level pattern), so that subsequent pattern optimization works as intended.
    - Please add tests (in `tests/test_transpiler.py` see also `tests/test_pattern.py`) and make sure the pattern simulation results are as expected even after various forms of pattern optimizations (see `tests/test_pattern.py`). 

Please see our [contribution guide](https://github.com/TeamGraphix/graphix/blob/master/CONTRIBUTING.md) before starting, and do not forget to comment on this issue to show your interest. We are always happy to answer questions, so do not hesitate to ask on this page.

Thank you for your interest, and good luck!

### PR DESCRIPTION
**Pull request by king-p3nguin**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- format added code and tests by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
 
Then, please fill in below:

**Context (if applicable):**

**Description of the change:**

**Related issue:**
- Resolves #49 

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

File: graphix/transpiler.py [hunk 1]
```diff
@@ -941,6 +981,283 @@ def _rzz_command_opt(self, control_node, target_node, ancilla, angle):
...
+        seq.append(["E", (ancilla[3], ancilla[4])])
+        seq.append(["E", (ancilla[4], control_node2)])
+        seq.append(["E", (control_node1, control_node2)])
+        seq.append(["M", ancilla[2], "XY", 0.25, [], [], 6])
+        seq.append(["M", ancilla[3], "XY", -0.25, [], [], 0])
+        seq.append(["M", ancilla[4], "XY", 0.25, [], [], 6])
+        seq.append(["M", ancilla[1], "XY", -0.25, [], [], 6])
+        seq.append(["M", ancilla[0], "XY", -0.25, [], [], 6])
+        seq.append(["M", ancilla[5], "XY", 0.25, [], [], 6])
+        # seq.append(["C", control_node1, 6]) # TODO: does not work?
```
Comment (king-p3nguin) [inline graphix/transpiler.py:1255]: @shinich1 
After running `pattern.perform_pauli_measurements()`, I got a series of commands, and there is a command which includes `"C"`.
Running the following code raises `TypeError: CommandNode.__init__() got an unexpected keyword argument 'vop'` and I am not sure how to add `"C"` command in the measurement pattern.
```python
from graphix.transpiler import Circuit
import numpy as np

circuit = Circuit(3)
# prepare |000>
circuit.h(0)
circuit.h(1)
circuit.h(2)
# prepare |111>
circuit.x(0)
circuit.x(1)
circuit.x(2)

circuit.ccx(0, 1, 2)
pattern = circuit.transpile(opt = True)

pattern.standardize()
pattern.shift_signals()
pattern.perform_pauli_measurements()
pattern.minimize_space()

from qiskit.quantum_info import Statevector
ket = pattern.simulate_pattern(backend='statevector').flatten()
ket = Statevector(ket)
display(ket.draw(output='latex'))
```
Traceback of the error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 17
     14 circuit.ccx(0, 1, 2)
     15 pattern = circuit.transpile(opt = True)
---> 17 pattern.standardize()
     18 pattern.shift_signals()
     19 pattern.perform_pauli_measurements()

File [~/graphix/graphix/pattern.py:246](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/kazuki_main/graphix/~/graphix/graphix/pattern.py:246), in Pattern.standardize(self, method)
    234 """Executes standardization of the pattern.
    235 'standard' pattern is one where commands are sorted in the order of
    236 'N', 'E', 'M' and then byproduct commands ('X' and 'Z').
   (...)
    243     defaults to 'local'
    244 """
    245 if method == "local":
--> 246     localpattern = self.get_local_pattern()
    247     localpattern.standardize()
    248     self.seq = localpattern.get_pattern().seq

File [~/graphix/graphix/pattern.py:229](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/kazuki_main/graphix/~/graphix/graphix/pattern.py:229), in Pattern.get_local_pattern(self)
    227     if index in self.output_nodes:
    228         node_prop[index]["output"] = True
--> 229     node = CommandNode(index, **node_prop[index])
    230     nodes[index] = node
    231 return LocalPattern(nodes, self.output_nodes, morder)

TypeError: CommandNode.__init__() got an unexpected keyword argument 'vop'
```

**Comment by shinich1** [comment]
@king-p3nguin hi, just want to see how it's going as UnitaryHack finishes early next week.

**Comment by king-p3nguin** [comment]
@shinich1 Thank you for the reminder. I am a bit busy and unsure I can meet the deadline, but I will work on this issue even if the UnitaryHack is finished.

**Comment by natestemen** [comment]
Nice work with this hack @king-p3nguin! Just a reminder that PRs are eligible for unitaryHACK bounties as long as the issue is closed before June 20th! The hackathon finishes EOD June 13th [AoE](https://time.is/Anywhere_on_Earth), and as long as the PR is up by this deadline, it is eligible! We understand it can take some time to get reviewed and make changes.

File: graphix/transpiler.py (hunk 1 above)
Comment (king-p3nguin) [reply graphix/transpiler.py:1255]: Hi, I would like to know how to skip specific Pauli measurements. Is it sufficient to simply add
```
if cmd[1] in nodes_skipped:
    continue
```
in the beginning of the following loop?

https://github.com/TeamGraphix/graphix/blob/de8692251d45947e42c2b57ae580d387c83b254b/graphix/pattern.py#L1633-L1649

Comment (shinich1) [reply graphix/transpiler.py:1255]: @king-p3nguin  Hi - yes, I think that should work.

Comment (king-p3nguin) [reply graphix/transpiler.py]: Hmm... I tried this method, but it is not working...
I skipped the 29th node Pauli measurement and got the following commands for opt, and implemented them as so, but the outcome is not correct.
```
N, node = 0
N, node = 1
N, node = 5
N, node = 9
N, node = 13
N, node = 21
N, node = 25
N, node = 28
N, node = 29
N, node = 30
E, nodes = (0, 25)
E, nodes = (1, 9)
E, nodes = (1, 25)
E, nodes = (5, 9)
E, nodes = (5, 25)
E, nodes = (9, 21)
E, nodes = (9, 29)
E, nodes = (13, 25)
E, nodes = (25, 29)
E, nodes = (29, 30)
M, node = 5, plane = XY, angle(pi) = 0.25, s-domain = [4, 2], t_domain = [], Clifford index = 6
M, node = 9, plane = XY, angle(pi) = -0.25, s-domain = [8, 6, 4, 2], t_domain = [], Clifford index = 0
M, node = 1, plane = XY, angle(pi) = -0.25, s-domain = [], t_domain = [], Clifford index = 6
M, node = 13, plane = XY, angle(pi) = 0.25, s-domain = [12, 10, 8, 6, 4, 2], t_domain = [], Clifford index = 6
M, node = 0, plane = XY, angle(pi) = -0.25, s-domain = [], t_domain = [], Clifford index = 6
M, node = 21, plane = XY, angle(pi) = 0.25, s-domain = [20, 14], t_domain = [], Clifford index = 6
M, node = 29, plane = XY, angle(pi) = 0, s-domain = [], t_domain = [], Clifford index = 0
X byproduct, node = 25, domain = [24]
X byproduct, node = 30, domain = [14, 20, 24, 26, 29]
X byproduct, node = 28, domain = [3, 5, 7, 9, 11, 13, 17, 19, 23]
Z byproduct, node = 25, domain = [0, 9, 11, 13, 17, 21, 27]
Z byproduct, node = 30, domain = [1, 5, 7, 9, 11, 15, 21, 27]
Z byproduct, node = 28, domain = [2, 4, 6, 8, 10, 12, 16, 18, 22]
```

Comment (shinich1) [reply graphix/transpiler.py]: hmm, in such a case, I recommend to leave input qubits unmeasured as well. how would that go?

Comment (king-p3nguin) [reply graphix/transpiler.py]: I think there is still some bug in the transpiler. The following code gave $\ket{010}$, which is incorrect.
```
from graphix.transpiler import Circuit
import numpy as np
circuit = Circuit(3)
# prepare |000>
circuit.h(0)
circuit.h(1)
circuit.h(2)
# prepare |011>
circuit.x(1)
circuit.x(2)
# toffoli
circuit.h(2)
circuit.cnot(1, 2)
circuit.rz(2, -np.pi/4)
circuit.cnot(0, 2)
circuit.rz(2, np.pi/4)
circuit.cnot(1, 2)
circuit.rz(1, np.pi/4)
circuit.rz(2, -np.pi/4)
circuit.cnot(0, 2)
circuit.cnot(0, 1)
circuit.rz(2, np.pi/4)
circuit.rz(0, np.pi/4)
circuit.rz(1, -np.pi/4)
circuit.h(2)
circuit.cnot(0, 1)
# transpile
pattern = circuit.transpile()
pattern.standardize()
pattern.shift_signals()
pattern.perform_pauli_measurements()
pattern.minimize_space()
# draw with latex
from qiskit.quantum_info import Statevector
ket = pattern.simulate_pattern(backend='statevector').flatten()
ket = Statevector(ket)
ket.draw(output='latex')
```

Comment (shinich1) [reply graphix/transpiler.py]: let me take a look

Comment (shinich1) [reply graphix/transpiler.py]: Thanks for waiting, the bug is fixed, the patch is on [this branch ](https://github.com/TeamGraphix/graphix/tree/62-bug-pattern-optimization-result-in-wrong-output-state-in-some-cases) and will be merged very soon (#63).

**Comment by shinich1** [comment]
@king-p3nguin Good that it works now! could you modify test cases to include ccz gate (perhaps add to random_circuit.py for now), and also add lines to docs (.rst files) where necessary?
also - perhaps you can use `-l 120` option for black to save up on line numbers in transpilar.py (see above)?

**Comment by king-p3nguin** [comment]
@shinich1 
Okay, I think I can do that. Unfortunately, there still seems to be some issue in the ccx commands...
I thought simply replacing the commands from the original pattern with each variable would work, but it did not. I wrote the following code to replace the commands. Maybe you can spot the bug.

```python
from graphix.transpiler import Circuit
import numpy as np

circuit = Circuit(3)
# toffoli
circuit.h(2)
circuit.cnot(1, 2)
circuit.rz(2, -np.pi / 4)
circuit.cnot(0, 2)
circuit.rz(2, np.pi / 4)
circuit.cnot(1, 2)
circuit.rz(1, np.pi / 4)
circuit.rz(2, -np.pi / 4)
circuit.cnot(0, 2)
circuit.cnot(0, 1)
circuit.rz(2, np.pi / 4)
circuit.rz(0, np.pi / 4)
circuit.rz(1, -np.pi / 4)
circuit.h(2)
circuit.cnot(0, 1)

# transpile
pattern = circuit.transpile()
pattern.standardize()
pattern.shift_signals()
pattern.perform_pauli_measurements()

dictonary = {  # for opt = True
    0: "ancilla[0]",
    1: "ancilla[1]",
    5: "ancilla[2]",
    9: "ancilla[3]",
    13: "ancilla[4]",
    21: "ancilla[5]",
    25: "control_node1",
    28: "target_node",
    29: "ancilla[6]",
    30: "control_node2",
}
# dictonary = { # for opt = False
#     0: "ancilla[0]",
#     1: "ancilla[1]",
#     2: "ancilla[2]",
#     3: "ancilla[3]",
#     4: "ancilla[4]",
#     5: "ancilla[5]",
#     6: "ancilla[6]",
#     7: "ancilla[7]",
#     8: "ancilla[8]",
#     9: "ancilla[9]",
#     10: "ancilla[10]",
#     11: "ancilla[11]",
#     12: "ancilla[12]",
#     13: "ancilla[13]",
#     14: "ancilla[14]",
#     15: "ancilla[15]",
#     16: "ancilla[16]",
#     17: "ancilla[17]",
#     18: "ancilla[18]",
#     19: "ancilla[19]",
#     20: "ancilla[20]",
#     21: "ancilla[21]",
#     22: "ancilla[22]",
#     23: "ancilla[23]",
#     24: "ancilla[24]",
#     26: "ancilla[25]",
#     27: "ancilla[26]",
#     29: "ancilla[27]",
#     25: "control_node1",
#     28: "target_node",
#     30: "control_node2",
# }
import numpy as np
for i in range(len(pattern.seq)):
    if pattern.seq[i][0] == "E":
        print(
            f'seq.append(["E", ({dictonary[pattern.seq[i][1][0]]}, {dictonary[pattern.seq[i][1][1]]})])'
        )
    elif pattern.seq[i][0] == "M":
        s_domain = pattern.seq[i][4]
        new_s_domain = []
        for sidx in s_domain:
            try:
                new_s_domain.append(dictonary[sidx])
            except KeyError:
                continue
        t_domain = pattern.seq[i][5]
        new_t_domain = []
        for tidx in t_domain:
            try:
                new_t_domain.append(dictonary[tidx])
            except KeyError:
                continue
        try:
            print(
                f'seq.append(["M", {dictonary[pattern.seq[i][1]]}, "{pattern.seq[i][2]}", {pattern.seq[i][3]}, [{", ".join(new_s_domain)}], [{", ".join(new_t_domain)}], {pattern.seq[i][6]}])'
            )
        except IndexError:
            print(
                f'seq.append(["M", {dictonary[pattern.seq[i][1]]}, "{pattern.seq[i][2]}", {pattern.seq[i][3]}, [{", ".join(new_s_domain)}], [{", ".join(new_t_domain)}]])'
            )
    elif pattern.seq[i][0] == "X":
        # remove duplicates
        _domain = np.array(pattern.seq[i][2])
        uind = np.unique(_domain)
        unique_domain = []
        for ind in uind:
            if np.mod(np.count_nonzero(_domain == ind), 2) == 1:
                unique_domain.append(ind)

        new_domain = []
        for idx in unique_domain:
            try:
                new_domain.append(dictonary[idx])
            except KeyError:
                continue
        if new_domain == []:
            continue
        print(
            f'seq.append(["X", {dictonary[pattern.seq[i][1]]}, [{", ".join(new_domain)}]])'
        )
    elif pattern.seq[i][0] == "Z":
        # remove duplicates
        _domain = np.array(pattern.seq[i][2])
        uind = np.unique(_domain)
        unique_domain = []
        for ind in uind:
            if np.mod(np.count_nonzero(_domain == ind), 2) == 1:
                unique_domain.append(ind)

        new_domain = []
        for idx in unique_domain:
            try:
                new_domain.append(dictonary[idx])
            except KeyError:
                continue
        if new_domain == []:
            continue
        print(
            f'seq.append(["Z", {dictonary[pattern.seq[i][1]]}, [{", ".join(new_domain)}]])'
        )
```

**Comment by king-p3nguin** [comment]
Also, I was a bit confused about how the "X" and "Z" commands work. `pattern.print_pattern()` seems to remove duplicate nodes in the list, but `StatevectorBackend.correct_byproduct(cmd)` does not?

**Comment by shinich1** [comment]
https://github.com/TeamGraphix/graphix/pull/56#issuecomment-1626050694

What's this code supposed to do? and what exactly is the issue? could you elaborate?

**Comment by shinich1** [comment]
> Also, I was a bit confused about how the "X" and "Z" commands work. `pattern.print_pattern()` seems to remove duplicate nodes in the list, but `StatevectorBackend.correct_byproduct(cmd)` does not?

the results are the same, we are thinking about a way to make sure we simplify signal lists in an efficient way, not just when printing

**Comment by king-p3nguin** [comment]
> [#56 (comment)](https://github.com/TeamGraphix/graphix/pull/56#issuecomment-1626050694)
> 
> What's this code supposed to do? and what exactly is the issue? could you elaborate?

Sorry, I did not explain it enough. What I was doing is that I generated the commands from a circuit that has only one Toffoli gate, and replaced the node number in the commands with `ancilla[i]`, `control_node1`, `control_node2`, and `target_node` so that I can simply copy & paste directly into the code.

**Comment by shinich1** [comment]
I see, it might be worth looking into graph state for toffoli gate and check their LC form after pauli measurement, e.g. look for isolated nodes. Just to check, have you modified `pattern.perform_pauli_measurements()` so it leave input qubits unmeasured? 
you may need to find alternative way and perhaps try out ZX calculus to get correct pattern fragment - rather than copy and paste.

**Comment by king-p3nguin** [comment]
Yes, I also tried skipping the measurement of input qubits and the results were the same.
I'm kind of stuck with this method, so will check ZX calculus.

**Comment by king-p3nguin** [comment]
By using PyZX and `graphix.generator.generate_from_graph()`, I was able to create commands for ccx gate (only for `opt = False`).
Commands I just created from `pattern.perform_pauli_measurements(leave_input=True)` does not seem to be working, so I will work on it later. (I might be missing something)

It took me a while to figure out how to convert a ZX-diagram into graph state. The process was more complicated than I expected. It might be a good idea to add to the example how to create a measurement pattern from a ZX-diagram created by pyzx. (It would also be a good use case for `graphix.generator.generate_from_graph()`)

**Comment by king-p3nguin** [comment]
I was not able to reduce the nodes a lot for `opt=True`, but it works for now.

@shinich1 Will you review this PR?

File: tests/test_tnsim.py [hunk 2]
```diff
@@ -302,6 +302,27 @@ def test_cnot(self):
         value2 = tn_mbqc.expectation_value(random_op2, [0, 1])
         np.testing.assert_almost_equal(value1, value2)
 
+    def test_ccx(self):
```
Comment (shinich1) [inline tests/test_tnsim.py:305]: I think it's best to test for reasonably randomized states (c.f. see earlier discussions of this PR and related PRs). Can you add randomizing circuit before the circuit here?

File: graphix/transpiler.py [hunk 3]
```diff
@@ -945,6 +985,238 @@ def _rzz_command_opt(self, control_node, target_node, ancilla, angle):
...
+        seq.append(["X", ancilla[17], [ancilla[14], ancilla[16]]])
+        seq.append(["X", ancilla[15], [ancilla[9], ancilla[11]]])
+        seq.append(["X", ancilla[13], [ancilla[0], ancilla[2], ancilla[5], ancilla[7], ancilla[12]]])
+        seq.append(["Z", ancilla[17], [ancilla[4], ancilla[5], ancilla[7], ancilla[10], control_node1]])
+        seq.append(["Z", ancilla[15], [control_node2, ancilla[2], ancilla[5], ancilla[10]]])
+        seq.append(["Z", ancilla[13], [ancilla[1], ancilla[3], ancilla[6], ancilla[8], target_node]])
+        return ancilla[17], ancilla[15], ancilla[13], seq
+
+    @classmethod
+    def _ccx_command_opt(self, control_node1, control_node2, target_node, ancilla):
```
Comment (shinich1) [inline graphix/transpiler.py:1114]: how did you get this opt version?

Comment (king-p3nguin) [reply graphix/transpiler.py:1114]: I converted the quantum circuit to a ZX diagram and ran `id_simp` and `spider_simp` on the diagram. I tried other simplifications like `clifford_simp` and `full_reduce` but was not able to generate a pattern from the diagram.

**Comment by shinich1** [comment]
@king-p3nguin found a better pattern, please take a look. what worked was to manually perform Pauli measurements of the pattern using graph state sim while making sure no local Hadamard remain on either input or output nodes (because these cannot be commuted through next or previous commands), sometimes interleaving with equivalent_fill_nodes of the graph state simulator. 

the script I used (works also after perform_pauli_measurements)
```python
circuit = Circuit(3)
circuit.ccx(0, 1, 2)
pattern = circuit.transpile()
pattern.standardize()
pattern.shift_signals()
pattern.draw_graph()
nodes, edges = pattern.get_graph()
vops = pattern.get_vops()
print(nodes)
print(edges)
from graphix import GraphState
nodes2 = [i for i in range(21)]
mapping = {nodes[i]: nodes2[i] for i in range(21)}
edges2 = [(mapping[i], mapping[j]) for i,j in edges]
vop2 = dict()
for i in nodes:
    if i in vops.keys():
        vop2[mapping[i]] = vops[i]
    else:
        vop2[mapping[i]] = 0
g=GraphState(nodes=nodes2, edges=edges2, vops=vop2)

results = dict()
results[3] = g.measure_x(3)
g.equivalent_fill_node(2)
results[4] = g.measure_x(4)
results[6] = g.measure_x(6)
results[12] = g.measure_x(12)
g.equivalent_fill_node(1)
results[11] = g.measure_x(11)
results[19] = g.measure_x(19)
results[9] = g.measure_x(9)
# 17, 14
g.draw()
print(results)

```

File: graphix/transpiler.py [hunk 4]
```diff
@@ -602,6 +675,8 @@ def _move_byproduct_to_right(self):
                 continue
             if self._instr[target + 1][0] == "CNOT":
```
Comment (shinich1) [inline graphix/transpiler.py:676]: we'll need to add _commute_with_ccx, to be complete. but I am thinking of discontinuing standardize_and_tranpile, so we might skip that for now

Comment (shinich1) [reply graphix/transpiler.py:676]: @masa10-f shall we discontinue `standardise_and_transpile`? I think there isn't much of benefit (#25).

**Comment by shinich1** [comment]
@king-p3nguin would you be happy to add type hints?

**Comment by shinich1** [comment]
@king-p3nguin looks great, thanks! if it's ready on your side, please squash and merge.

**Pull request merged**